pkg/agent/nodeportlocal/k8s/annotations.go

```diff
@@ -45,106 +46,84 @@ func toJSON(serialize interface{}) string {
-			PodPort:  containerPort,
-			NodeIP:   nodeIP,
-			NodePort: nodePort,
-		}}
 	}
-
-	current[NPLAnnotationKey] = toJSON(annotations)
-	pod.Annotations = current
-}
-
-func removePodAnnotation(pod *corev1.Pod) {
-	klog.V(2).Infof("Removing entire NodePortLocal annotation from Pod: %s/%s", pod.Namespace, pod.Name)
-	delete(pod.Annotations, NPLAnnotationKey)
-}
-
-func removeFromPodAnnotation(pod *corev1.Pod, containerPort int) {
-	var err error
-	current := pod.Annotations
-
-	klog.V(2).Infof("Removing NodePortLocal annotation from Pod: %s/%s\tport: %v", pod.Namespace, pod.Name, containerPort)
-	var annotations []NPLAnnotation
-	if err = json.Unmarshal([]byte(current[NPLAnnotationKey]), &annotations); err != nil {
-		klog.Warningf("Unable to unmarshal NodePortLocal annotation: %v", current[NPLAnnotationKey])
-		return
+	if isNodePortInAnnotation(nplAnnotations, nodePort, containerPort) {
+		// no updates required to the Pod
+		return false
 	}
+	return true
+}
 
+func removeFromNPLAnnotation(annotations []NPLAnnotation, containerPort int) []NPLAnnotation {
 	for i, ann := range annotations {
 		if ann.PodPort == containerPort {
 			annotations = append(annotations[:i], annotations[i+1:]...)
 			break
 		}
 	}
+	return annotations
+}
 
-	current[NPLAnnotationKey] = toJSON(annotations)
-	pod.Annotations = current
+func (c *NPLController) updatePodNPLAnnotation(pod *corev1.Pod, annotations []NPLAnnotation) error {
+	if err := patchPod(annotations, *pod, c.kubeClient); err != nil {
+		klog.Warningf("Unable to patch NPL annotations for Pod %s/%s: %s", pod.Namespace, pod.Name, err.Error())
+	}
+	klog.V(2).Infof("Successfully updated annotation for Pod %s/%s", pod.Namespace, pod.Name)
+	return nil
 }
 
-// RemoveNPLAnnotationFromPods removes npl annotations from all Pods.
-func (c *NPLController) RemoveNPLAnnotationFromPods() {
-	nodeName, err := env.GetNodeName()
-	if err != nil {
-		klog.Warningf("Failed to get Node's name, NodePortLocal annotation cannot be removed for Pods scheduled to this Node")
-		return
+func patchPod(value []NPLAnnotation, pod corev1.Pod, kubeClient clientset.Interface) error {
+	type Metadata struct {
+		Annotations map[string]*string `json:"annotations"`
 	}
-	podList, err := c.kubeClient.CoreV1().Pods(metav1.NamespaceAll).List(context.TODO(), metav1.ListOptions{
-		FieldSelector:   "spec.nodeName=" + nodeName,
-		ResourceVersion: "0",
-	})
-	if err != nil {
-		klog.Warningf("Unable to list Pods, err: %v", err)
-		return
+	type PatchData struct {
+		Meta Metadata `json:"metadata"`
 	}
-	for i, pod := range podList.Items {
-		if _, exists := pod.Annotations[NPLAnnotationKey]; !exists {
-			continue
-		}
-		removePodAnnotation(&podList.Items[i])
-		c.updatePodAnnotation(&podList.Items[i])
+
+	payloadValue := make(map[string]*string)
+	if len(value) > 0 {
+		valueStr := string(toJSON(value))
+		payloadValue[NPLAnnotationKey] = &valueStr
+	} else {
+		payloadValue[NPLAnnotationKey] = nil
+	}
+	payload := PatchData{
```

Comment:
You can use anonymous structs here too.

Reply:
Since we are not using struct members anywhere, just changed the structs to maps instead.